newlib/libc/stdlib/nano-mallocr.c

```diff
@@ -55,33 +56,15 @@
 #define assert(x) ((void)0)
 #endif
 
-#ifndef MAX
-#define MAX(a,b) ((a) >= (b) ? (a) : (b))
-#endif
-
-#if __SIZEOF_POINTER__ == __SIZEOF_LONG__
-#define ALIGN_TO(size, align) \
-    (((size) + (align) -1L) & ~((align) -1L))
-#else
-#define ALIGN_TO(size, align) \
-    (((size) + (align) -1) & ~((align) -1))
-#endif
-
-#define ALIGN_PTR(ptr, align)	(void *) (uintptr_t) ALIGN_TO((uintptr_t) ptr, align)
-
-typedef struct {
-    char c;
-    union {
-	void *p;
-	double d;
-	long long ll;
-	size_t s;
-    } u;
+typedef union {
+    void *p;
+    double d;
+    long long ll;
+    size_t s;
 } align_chunk_t;
 
 typedef struct {
-    char c;
-    size_t s;
+    uintptr_t s;
```

Comment:
Why the change in type here? I have this vague memory of platforms where size_t is smaller than uintptr_t.

Reply:
That's exactly the type of platform I made this change for, for CHERI-enabled platforms size_t is half the size of pointers, but malloc still needs to return pointer-aligned values.

I believe it should be possible to use size_t instead of uintptr_t here, but having the header be less aligned than the malloc_chunk requires some additional non-trivial changes to the remaining code. I went for the simplest approach here which is to use uintptr_t which should be no-op change for all other platforms.
It's been a while since I made these changes, but here is what I wrote in the commit message back then:
> For CHERI size_t is half the size of uintptr_t (and thus pointers), so
using uinptr_t instead of size_t ensures there is no padding between
the malloc_chunk members `size` and `next`. This is required by the
remaining logic since otherwise the `next` pointer could become
unaligned if we only subtract MALLOC_HEAD (size_t) instead of a pointer
size quantity. Further changes would allow using a single size_t before
the actual allocation, but this minimally invasive change has no effect
on existing architectures while enabling the malloc test to pass for
CHERI purecap architectures.

Reply:
Oh, it's actually designed to let the header be allocated on a less strict alignment than the data. For instance, on a regular 32-bit platform, size_t is 32 bits but malloc wants to return blocks aligned on 64-byte boundaries for long long and double values. If you allocate 12 bytes, there's a nice 4-byte chunk after your block but before the next 64-bit boundary that the size can fit into.

Of course, on that platform, uintptr_t and size_t are both 32-bits, so there's no impact. With Cheri, we should be able to get the same result by using the smaller size_t type?

Reply:
That's a good point, it should only really matter for platforms where malloc alignment is > pointer size, and in that case size_t will be the same as uintptr_t, so I think using uintptr_t is still a no-op change?

Reply:
> That's a good point, it should only really matter for platforms where malloc alignment is
> pointer size, and in that case size_t will be the same as uintptr_t,
> so I think using uintptr_t is still a no-op change?

Most people would expect a size_t there. Could you please explain again in which scenario the type uintptr_t is better than size_t?

Reply:
sure, `uintptr_t` and `size_t` are generally the same underlying integer type. However, using `uintptr_t` instead of `size_t` is misleading about what this field holds -- it doesn't hold an integer representation of a pointer value, it holds the size of the allocation, and that is defined in the interface to `malloc` as `size_t`.

Is there any reason that you need to use `uintptr_t` instead of `size_t` for this field?